### PR TITLE
Improve overriding endpoint when not primary

### PR DESCRIPTION
**Pull request by designatednerd**
a) Cleaned up some code. 
b) The main point of this PR: 

This change is inspired by an issue I saw in code coverage while trying to bump up some coverage with Toshi's tests. 

If you have to hit an underlying endpoint (like Toshi's timestamp endpoint) to hit your target endpoint and pass in a status code, the error you'll get will actually be from the overridden underlying endpoint and not from the main endpoint. This basically means that even though you think you're doing at least some testing of error handling on your target endpoint, you're not - you're only testing the error handling on the overridden underlying endpoint (meaning, you're only testing that the timestamp endpoint handles errors properly, not  the actual thing you're targeting).

This updates behavior so that if an endpoint is overridden using `overrideEndpoint`, that call will still succeed and it is the targeted call which will fail. 

So basically before: 
-  User tries to test `getUsers`, which under the hood calls `getTimestamp`, passing in a failure status code and overriding the `timestamp` endpoint to provide a timestamp. 
- The call fails at `getTimestamp` and is returned with the failure status code. 
- None of the main call executes. 

And now: 
-  User tries to test `getUsers`, which under the hood calls `getTimestamp`, passing in a failure status code and overriding the `timestamp` endpoint to provide a timestamp. 
- The `getTimestamp` call succeeds and proceeds to the main call.
- The failure is then returned through the main call. 

Let me know if y'all would like further clarification on this - i think the code probably does a better job of explaining it than this novella. 😛

### CODE REVIEW

File: TeapotTests/MockTests.swift [hunk 1]
```diff
@@ -84,11 +84,34 @@ class MockTests: XCTestCase {
...
-                XCTAssertEqual(json!.dictionary!["overridden"] as! String, "value")
+                XCTAssertEqual(json?.dictionary?["overridden"] as? String, "value")
             case .failure(let error):
                 print(error)
                 XCTFail()
             }
         }
     }
+    
+    func testEndpointOverriddingThenHittingOtherEndpointWithError() {
```
Comment (elland) [inline TeapotTests/MockTests.swift:95]: In these cases, I always find it very useful to document the test with the expected json I'm sending/receiving. Makes both following the code and checking it easier.

File: TeapotTests/ResponseTests.swift [hunk 2]
```diff
@@ -42,8 +42,8 @@ class ResponseTests: XCTestCase {
         XCTAssertNotNil(json.data)
         XCTAssertNotNil(json.array)
         // internal data uses [] for dicts as well as arrays, so:
-        XCTAssertNotEqual(json.data!, data)
```
Comment (elland) [inline TeapotTests/ResponseTests.swift:45]: 💅🏽

File: TeapotTests/TeapotTests.swift [hunk 3]
```diff
@@ -144,21 +140,24 @@ class TeapotTests: XCTestCase {
...
-        // http://icons.iconarchive.com 
-        // /icons/martz90/circle/512/app-draw-icon.png
         let expectation = self.expectation(description: "GetImage")
         let url = URL(string: "http://icons.iconarchive.com/icons/martz90/circle/512/app-draw-icon.png")!
 
         Teapot(baseURL: url).get() { (result: NetworkImageResult) in
             switch result {
             case .success(let image, let response):
-                let localImage = Bundle(for: TeapotTests.self).image(forResource: NSImage.Name(rawValue: "app-draw-icon"))!
+                guard let localImage = Bundle(for: TeapotTests.self).image(forResource: NSImage.Name(rawValue: "app-draw-icon")), let tiff = localImage.tiffRepresentation else {
```
Comment (elland) [inline TeapotTests/TeapotTests.swift]: I find having the local constant declared in the middle of a long line a bit hard to follow, compared to breaking it down line-by-line. What do you think?

**Review by elland — CHANGES_REQUESTED** [review]
Took me longer than it should to fully grasp what's going on but looks good.

One thing it's really missing tho is a nice README update with an example case. 💪🏽

File: TeapotTests/TeapotTests.swift (hunk 3 above)
Comment (designatednerd) [reply TeapotTests/TeapotTests.swift]: I agree but I was following our general coding standards for this. I'm definitely 👍  to break it up into two lines.

Comment (elland) [reply TeapotTests/TeapotTests.swift]: We have that in the guidelines? 🙈 PR time there I guess hehe

**Comment by designatednerd** [comment]
@Elland added a nice long explanation with examples for mocking.

**Review by elland — APPROVED** [review]
👏🏽
